surfer/viz.py

```diff
@@ -2813,18 +2936,40 @@ def add_overlay(self, old):
         return OverlayDisplay(self, array_id, pos, pos_bar, neg, neg_bar)
 
     @verbose
-    def add_data(self, array, mlab_plot, min, max, thresh, lut, colormap,
-                 alpha, colorbar, layer_id):
+    def add_data(self, array, fmin, fmid, fmax, thresh, lut, colormap, alpha,
+                 colorbar, layer_id, smooth_mat, magnitude, magnitude_max,
+                 scale_factor, vertices, vector_alpha):
         """Add data to the brain"""
         # Calculate initial data to plot
         if array.ndim == 1:
             array_plot = array
         elif array.ndim == 2:
             array_plot = array[:, 0]
+        elif array.ndim == 3:
+            assert array.shape[1] == 3  # should always be true
+            assert magnitude is not None
+            assert scale_factor is not None
+            array_plot = magnitude[:, 0]
         else:
-            raise ValueError("data has to be 1D or 2D")
+            raise ValueError("data has to be 1D, 2D, or 3D")
+        vector_values = array_plot
+        if smooth_mat is not None:
+            array_plot = smooth_mat * array_plot
 
-        array_id, pipe = self._add_scalar_data(mlab_plot)
+        # Copy and byteswap to deal with Mayavi bug
+        array_plot = _prepare_data(array_plot)
+
+        array_id, pipe = self._add_scalar_data(array_plot)
+        scale_factor_norm = None
+        if array.ndim == 3:
+            scale_factor_norm = scale_factor / magnitude_max
+            vectors = array[:, :, 0].copy()
+            glyphs = self._add_vector_data(
+                vectors, vector_values, fmin, fmid, fmax,
+                scale_factor_norm, vertices, vector_alpha, lut)
+        else:
+            glyphs = None
+        del scale_factor
```

Comment:
Same as above, here IIRC it was to prevent using something later that shouldn't be used